cpp/src/Ice/ConnectionI.cpp

```diff
@@ -3493,17 +3493,19 @@ Ice::ConnectionI::dispatchAll(
 Ice::ConnectionInfoPtr
 Ice::ConnectionI::initConnectionInfo() const
 {
-    // Called with _mutex locked.
-
     if (_state > StateNotInitialized && _info) // Update the connection information until it's initialized
     {
         return _info;
     }
 
+    bool incoming = !_connector;
+
+    // _adapter is set for an incoming connection until "finished"
     _info = _transceiver->getInfo(
-        _connector == nullptr,
-        _adapter ? _adapter->getName() : string{},
+        incoming,
+        incoming && _adapter ? _adapter->getName() : string{},
```

Comment:
Can we cache the adapter name in the connection, and always provide it for incoming connections?

Reply:
We could, but this seems very minor and hard to test. _adapter is usually not null for an incoming connection.

Reply:
Ok, the code is clear now that you remove the asserts.